### ISSUE TITLE
IPv6 Support

### ISSUE THREAD

**Issue by owendelong** [esp8266/Arduino#638]
Currently the entire ESP8266 implementation is IPv4 monastack. It would be nice to be able to choose (through compile-time directives or other mechanism) IPv4, IPv6, or IPv4v6 dual stack.

The lack of IPv6 support is already proving a hinderance in some environments where I work and I can only imagine that this will get worse as we go forward. The ARIN free pool is down to fewer than 256 /24 prefixes. Less than 65,000 total available IPv4 addresses for all of north america.
##

**Comment by far5893** [comment]
I suppose that it depend from expressif OS and not from arduino wrapper.

**Comment by owendelong** [comment]
Perhaps, but I'm betting if the Arduino community as a whole started pushing for this, Espressif would be more likely to listen than to just me. I've also put this forward with them as well.

**Comment by igrr** [comment]
For the record, they have already released lwIP source code for 1.3.0 SDK.

**Comment by mst78** [comment]
There seems to be IPV6 support in the espressif RTOS SDK1.3:
http://bbs.espressif.com/viewtopic.php?t=619#p4444

@Igrr: Is it possible to integrate IPV6 now? Do you have any plans of doing it?

**Comment by ccoenen** [comment]
I'm interested in IPv6 support, as well!

**Comment by Toshik** [comment]
Any news on IPv6?

**Comment by Skeen** [comment]
+1 for IPv6

**Comment by kaeferfreund** [comment]
+1

**Comment by igrr** [comment]
IPv6 support isn't present in the non-OS SDK, which is used by this project.
RTOS SDK, on the other hand, has IPv6. An Arduino core for the ESP8266/ESP32 based on RTOS SDK will be released as a separate project.

**Comment by mst78** [comment]
Igrr, do you have any idea when the first version of Arduino core for the ESP8266/ESP32 based on RTOS SDK will be available?
Have you already started working on it?

**Comment by igrr** [comment]
Yes, i was in fact working on things related to IPv6 support in ESP32 Arduino core yesterday :) 
I can't give you any concrete date unfortunately. It's going to be available within next two months, this is somewhat gated by the release of new SDK version.

For now, there is a publicly available ESP32 core by @me-no-dev. It has partial IPv6 support. LwIP stack has support for IPv6, but it is not complete (missing IPv6 support in DNS client, for instance). On the Arduino side of things, IPAddress class in this core also doesn't support IPv6 addresses.

**Comment by MyIgel** [comment]
@igrr Any updates?

**Comment by igrr** [comment]
Regarding ipv6 for the 8266 non-OS SDK, I don't have any updates.

**Comment by FlorianHeigl** [comment]
ping, ipv4 is deprecated!
I was just checking this to install a few dozen at a conference. it's unlikely they still allow ipv4.
is this something you do at work? can you tell your boss this actually matters?
If it's just out of your free time and energy, then don't worry! :)

**Comment by kohlsalem** [comment]
+1 and quite surprised to see a product so recent w/o ipv6

**Comment by jbtrystram** [comment]
+1 :)

**Comment by itavero** [comment]
Has anyone looked at integrating picoTCP?
I'm pretty sure they have had IPv6 support for a while now.
https://github.com/tass-belgium/picotcp

**Comment by MatthiasLohr** [comment]
Anything new about this topic? I' interested in IPv6, too...

**Comment by NiKiZe** [comment]
Sorry to bump this but, any update?

**Comment by ttytyper** [comment]
In case nobody has noticed, there's a $30 bounty on this issue: https://www.bountysource.com/issues/25224540-ipv6-support/backers

**Comment by FlorianHeigl** [comment]
Do you think the bounty will help? just wondering since there's no reply here ever since...
Would not mind throwing more at it. I bet most of us would do that.

**Comment by krmax44** [comment]
Has there been any progress since then? IPv6 is a very important thing in the IOT world.

**Comment by d-a-v** [comment]
Doable. At least tryable. Buy me time, here are some [hints](https://github.com/d-a-v/esp82xx-nonos-linklayer/issues/8#issuecomment-354570604). core-2.4.0 has first to happen though ;)

**Comment by devyte** [comment]
The precursor was migrating to lwip2, and that is in. Now it just needs to be stabilized. After that, the glue layer between the core and lwip needs some changes. After that, our socket layer needs to get support. After that, the rest of the core and libs.
It will take a (long) while, but now at least it's possible.

**Comment by FlorianHeigl** [comment]
Thank you @d-a-v and @devyte for giving outsiders some insight and also for having it on your mind, and for all that you did towards it. seems a lot of pieces need to be moved :)

**Comment by d-a-v** [comment]
#5136

**Comment by C0rn3j** [comment]
Just decided to check up on this for the first time and I see a PR 2 hours ago... great job!

**Comment by C0rn3j** [comment]
@d-a-v would love to test but failing to find the "Arduino application directory" on Arch Linux (Arduino 1.8.5). (Trying to follow this https://github.com/esp8266/Arduino#using-git-version )

Seems like the esp8266 package, when installed through the IDE, installs under ~/.arduino15, but libraries like IRremoteESP8266 install under ~/Arduino.

I've tried installing it under the already-installed esp8266 folder and that broke my setup, so that was a bad guess.

Would you have an idea where exactly this folder would be under Linux?

**Comment by herrold** [comment]
@C0rn3j Your question really seems like something to answer by reviewing how Arch has packaged it, and then to solve the 'wrong' directory issue for libraries with a symlink

still packaging into ~/.auduino15/ for the base IDE seems ... very stale

**Comment by C0rn3j** [comment]
@herrold Oh right, thanks, I read the package files and it's in /usr/share/arduino (I guess I could have just looked where the executable is pointing to).

@d-a-v I cloned the git master and pulled your changes as per your OP in the PR and I got this when trying to compile my project.

```
Build options changed, rebuilding all
In file included from /usr/share/arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include/lwip/memp.h:58:0,
                 from /usr/share/arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include/lwip/stats.h:43,
                 from /usr/share/arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include/lwip/netif.h:50,
                 from /usr/share/arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include/lwip/ip.h:46,
                 from /usr/share/arduino/hardware/esp8266com/esp8266/libraries/ESP8266WiFi/src/WiFiClient.cpp:37:
/usr/share/arduino/hardware/esp8266com/esp8266/tools/sdk/lwip2/include/lwip/priv/memp_priv.h:48:32: fatal error: lwip/priv/mem_priv.h: No such file or directory
 #include "lwip/priv/mem_priv.h"
                                ^
compilation terminated.
exit status 1
Error compiling for board NodeMCU 1.0 (ESP-12E Module).
```

**Comment by d-a-v** [comment]
@C0rn3j sorry about that. The PR is updated

**Comment by earlephilhower** [comment]
Closing as fixed by #5136

**Comment by thrust26** [comment]
I recently got a product which uses the ESP8266 for WiFi connect. The Router only supports IPv6 and immediately deauthenticates the connections from the ESP8266 (Reason-Code 3).

Any idea what's wrong here?

**Comment by d-a-v** [comment]
@thrust26 Please open a new issue with all the details you can provide by filling the issue template.

**Comment by thrust26** [comment]
After a bit more research I am pretty the firmware does not support IPv6. Ignore my post above.
